sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -209,20 +479,291 @@ node add_malloc_device(void *&data, size_t numBytes, const std::vector<node>& de
+for submitting a graph for execution, before a previous execution has been
+completed is backend specific. The runtime may throw an error.
+
+Parameters:
+
+* `graph` - Graph object to start recording commands to.
+
+When the queue is in the execution state, an `event` object used to synchronize
+graph submission with other command-groups is returned. Otherwise the queue is
+in the recording state, and a default event is returned.
+|===
+
+=== Thread Safety
+
+The new functions in this extension are thread-safe, the same as member
+functions of classes in the base SYCL specification. If user code does
+not perform synchronisation between two threads accessing the same queue,
+there is no strong ordering between events on that queue, and the kernel
+submissions, recording and finalization will happen in an undefined order.
+
+In particular, when one thread ends recording on a queue while another
+thread is submitting work, which kernels will be part of the subsequent
+graph is undefined. If user code enforces a total order on the queue
+events, then the behaviour is well-defined, and will match the observable
+total order.
+
+The returned value from the `info::queue::state` should be considered
+immediately stale in multi-threaded usage, as another thread could have
+preemptively changed the state of the queue.
+
+=== Error Handling
+
+Errors are reported through exceptions, as usual in the SYCL API. For new APIs,
+submitting a graph for execution can generate unspecified asynchronous errors,
+while `command_graph::finalize()` may throw unspecified synchronous exceptions.
+Synchronous exception errors codes are defined for both
+`queue::begin_recording()` and `command_graph::update()`.
+
+When a queue is in recording mode asynchronous exceptions will not be
+generated, as no device execution is occuring. Synchronous errors specified as
+being thrown in the default queue executing state, will still be thrown when a
+queue is in the recording state.
+
+The `queue::begin_recording` and `queue::end_recording` entry-points return a
+`bool` value informing the user whether a state change occurred. False is
+returned rather than throwing an exception when state isn't changed. This design
+is because the queue is already in the state the user desires, so if the
+function threw an exception in this case, the application would likely swallow
+it and then proceed.
+
+While a queue is in the recording state, methods performed on that queue which
+are not command submissions behave as normal. This includes waits, throws, and
+queries on the queue. These are all ignored by the graph system, as opposed to
+throwing an exception when in queue recording mode. This is because otherwise
+there would be no thread safe way for a user to check they could call these
+functions without throwing, as a query about the state of the queue may be
+immediately stale.
+
+* TODO - error on add_node while being recorded to a queue? or queue recording a
```

Comment:
I haven't added any specification wording on what happens a user mixes graph building mechanisms.

If a modifiable graph is being recorded to by a queue, and then the user uses explicit APIs to add nodes to the same graph before the recording has finished, is that okay? We should specify if that's valid or not.

Another possible scenario is if a modifiable graph contains nodes which have been explicitly added, and then the user goes and records more commands to it from a queue. Is that okay?

Reply:
Now tracked in https://github.com/reble/llvm/issues/11